src/packages/blits/src/components/Badge/Badge.ts

```diff
@@ -0,0 +1,162 @@
+  iconColor?: string;
+  iconSrc?: string;
+};
+
+type BadgeState = {
+  padding: Padding;
+  font: string;
+  fontSize: number;
+  fontWeight: number;
+  lineHeight: number;
+  gap: number;
+};
+
+const Badge = Blits.Component('Badge', {
+  components: {
+    Icon
+  },
+  props: [
+    {
+      key: 'title',
+      required: false,
+      cast: (v?: string): string | undefined => v
+    },
+    {
+      key: 'iconSrc',
+      required: false,
+      cast: (v?: string): string | undefined => v
+    },
+    {
+      key: 'iconColor',
+      required: false,
+      cast: (v?: string): string | undefined => v
+    },
+    {
+      key: 'iconAlign',
+      default: 'left',
+      required: false,
+      cast: (v: unknown): IconAlign | undefined => {
+        if (v === undefined || isType<IconAlign>(v, IconAlignments)) return v;
+        throw new Error(`Invalid icon alignment '${v}'`);
+      }
+    },
+    {
+      key: 'tone',
+      default: 'neutral',
+      required: false,
+      cast: (v: string): Tone => {
+        if (isValidTone(v)) return v;
+        throw new Error(`Invalid tone '${v}'`);
+      }
+    }
+  ],
+  // iconSrc defaulted with empty string to prevent blits warning for required field - would be better to conditionally render the icon if possible
+  template: `
+    <Element
+      :w="$containerSize.w"
+      :h="$containerSize.h"
+      :color="$containerColor"
+      :effects="[$shader('radius', {radius: $border.radius}), $shader('border', {width: $border.width, color: $border.color})]"
+    >
+      <Icon
+        :x="$iconLeft"
+        :y="$padding[0]"
+        :width="$icon.width"
+        :height="$icon.height"
+        :show="$icon.show"
+        :iconSrc="$icon.src ?? ''"
+        :iconColor="$icon.color"
+      />
+      <Text
+        :x="$textLeft"
+        :y="$padding[0]"
+        :content="$title"
+        :lineheight="$lineHeight"
+        :size="$fontSize"
+        :font="$font"
+        :color="$textColor"
+      />
+    </Element>
+  `,
+  state(): BadgeState {
+    return {
+      padding: styles.Container.base.padding,
+      font: styles.Text.base.fontFamily,
+      fontSize: styles.Text.base.fontSize,
+      fontWeight: styles.Text.base.fontWeight,
+      lineHeight: styles.Text.base.lineHeight,
+      gap: styles.Container.base.gap
+    };
+  },
+  computed: {
+    iconLeft(): number {
+      return (
+        this.padding[3] +
+        (this.icon.show && this.iconAlign === 'right' && this.textWidth && this.textWidth + this.gap)
+      );
+    },
+    textLeft(): number {
+      return this.padding[3] + (this.icon.show && this.iconAlign !== 'right' && this.icon.width + this.gap);
+    },
+    textWidth(): number {
+      // calculate width of text using dummy canvas element - hopefully blits will get flex soon so we dont need to do this
+      const canvas = document.createElement('canvas');
+      const ctx = canvas.getContext('2d');
+      ctx!.font = `${this.fontSize}px ${this.font}`;
```

Comment:
I haven't seen `!.` before. What does it do?

Reply:
i wasn't familiar with it either, looks like it's a typescript [non-null assertion operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator). pretty neat

Reply:
the primary difference between this and optional chaining seem to be:
- **optional chaining** is JS and tells the runtime "if this is undefined, bail out of the lookup"
- **non-null assertion** is strictly TS and tells the compiler "this definitely won't be null/undefined so don't worry about that"

Reply:
Yeh I've found `!` and `?` really useful in the past, in this case thought it was safe enough given it's creating the canvas there and then, only case it might be null is if a browser doesn't support it but in that case I don't think it'd be able to run lightning anyway